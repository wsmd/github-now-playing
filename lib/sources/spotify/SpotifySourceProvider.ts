import osascript from 'osascript-tag';
import { SourceProvider } from '../../SourceProvider';
import { NowPlayingTrack } from '../../types';

export class SpotifySourceProvider extends SourceProvider {
  protected getNowPlaying(): Promise<NowPlayingTrack | null> {
    return osascript.jxa<NowPlayingTrack | null>({ parse: true })`
      const spotify = Application('Spotify');

      if (!spotify.running()) {
        return null;
      }

      if (spotify.playerState() !== 'playing') {
        return null;
      }

      return {
        artist: spotify.currentTrack.artist(),
        title: spotify.currentTrack.name(),
      }
    `;
  }
}
