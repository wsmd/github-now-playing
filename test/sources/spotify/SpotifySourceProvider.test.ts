import osascript from 'osascript-tag';
import { SpotifySourceProvider } from 'lib/sources/spotify/SpotifySourceProvider';

beforeEach(jest.useFakeTimers);
afterEach(jest.clearAllTimers);

describe('SpotifySourceProvider', () => {
  it('calls osascript.jxa with correct script', () => {
    const runScriptFn = jest.fn();
    jest.spyOn(osascript, 'jxa').mockImplementationOnce(() => runScriptFn);
    const source = new SpotifySourceProvider({ updateFrequency: 10000 });
    source.listen();
    expect(runScriptFn.mock.calls[0]).toMatchSnapshot();
  });

  it('reports the track via osascript', async () => {
    const expectedTrack = { artist: 'UFO', title: 'Space Child' };

    // osascript.jxa does returns a track
    jest.spyOn(osascript, 'jxa').mockImplementation(() => async () => expectedTrack);

    const onChangeFn = jest.fn();

    const source = new SpotifySourceProvider({ updateFrequency: 1 });
    source.on(SpotifySourceProvider.Events.TrackChanged, onChangeFn);
    source.listen();

    await new Promise(process.nextTick);
    expect(onChangeFn).toHaveBeenCalledWith(expectedTrack);
  });

  it('does not report a track if nothing is currently playing', async () => {
    // osascript.jxa does not return a track
    jest.spyOn(osascript, 'jxa').mockImplementation(() => async () => null);

    const onChangeFn = jest.fn();
    const source = new SpotifySourceProvider({ updateFrequency: 1 });
    source.on(SpotifySourceProvider.Events.TrackChanged, onChangeFn);
    source.listen();

    await new Promise(process.nextTick);
    expect(onChangeFn).not.toHaveBeenCalled();
  });
});
