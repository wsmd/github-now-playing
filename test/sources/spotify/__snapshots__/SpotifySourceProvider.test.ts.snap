// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifySourceProvider calls osascript.jxa with correct script 1`] = `
Array [
  Array [
    "
      const spotify = Application('Spotify');

      if (!spotify.running()) {
        return null;
      }

      if (spotify.playerState() !== 'playing') {
        return null;
      }

      return {
        artist: spotify.currentTrack.artist(),
        title: spotify.currentTrack.name(),
      }
    ",
  ],
]
`;
